rdflib/graph.py

```diff
@@ -670,35 +670,125 @@ def set(self, triple):
-        for s, p, o in self.triples((None, predicate, None)):
-            yield s, o
-
-    def predicate_objects(self, subject=None):
-        """A generator of (predicate, object) tuples for the given subject"""
-        for s, p, o in self.triples((subject, None, None)):
-            yield p, o
+    def subjects(self, predicate=None, object=None, unique=False) -> Iterable[Node]:
+        """A generator of (optionally unique) subjects with the given
+        predicate and object"""
+        if not unique:
+            for s, p, o in self.triples((None, predicate, object)):
+                yield s
+        else:
+            subs = set()
+            for s, p, o in self.triples((None, predicate, object)):
+                if s not in subs:
+                    yield s
+                    try:
+                        subs.add(s)
+                    except MemoryError as e:
+                        logger.error(
+                            f"{e}. Consider not setting parameter 'unique' to True"
+                        )
+                        raise
+
+    def predicates(self, subject=None, object=None, unique=False) -> Iterable[Node]:
+        """A generator of (optionally unique) predicates with the given
+        subject and object"""
+        if not unique:
+            for s, p, o in self.triples((subject, None, object)):
+                yield p
+        else:
+            preds = set()
+            for s, p, o in self.triples((subject, None, object)):
+                if p not in preds:
+                    yield p
+                    try:
+                        preds.add(p)
+                    except MemoryError as e:
+                        logger.error(
+                            f"{e}. Consider not setting parameter 'unique' to True"
+                        )
+                        raise
+
+    def objects(self, subject=None, predicate=None, unique=False) -> Iterable[Node]:
+        """A generator of (optionally unique) objects with the given
+        subject and predicate"""
+        if not unique:
+            for s, p, o in self.triples((subject, predicate, None)):
+                yield o
+        else:
+            objs = set()
+            for s, p, o in self.triples((subject, predicate, None)):
+                if o not in objs:
+                    yield o
+                    try:
+                        objs.add(o)
+                    except MemoryError as e:
+                        logger.error(
+                            f"{e}. Consider not setting parameter 'unique' to True"
+                        )
+                        raise
+
+    def subject_predicates(
+        self, obj=None, unique=False
+    ) -> Generator[Tuple[Node, Node], None, None]:
+        """A generator of (optionally unique) (subject, predicate) tuples
+        for the given object"""
+        if not unique:
+            for s, p, o in self.triples((None, None, obj)):
```

Comment:
To match earlier change

```suggestion
            for s, p, o in self.triples((None, None, object)):
```